_posts/2023-10-04-#28107.md

```diff
@@ -7,17 +7,59 @@ authors: [dergoegge]
+
 ## Questions
+
+### Concept
+
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. What does it mean for a transaction identifier to be type-safe? Why is that important or helpful? Are there any downsides?
+
+1. Why is it better to enforce types at compile-time instead of at run-time?
+
+1. Conceptually, when writing new code that requires referring to transactions, when should you use `txid` and when should you use `wtxid`? Can you point to any examples in the code where using one instead of the other could be very bad? (1 point per example)
+
+1. In which concrete way(s) could using `transaction_identifier` instead of `uint256` help find existing bugs or prevent the introduction of new ones? On the other hand, could this change introduce new bugs?
+
+1. SegWit activated at block height 481,824. Do transactions in blocks < 481,824 also have a `wtxid`? For example, what about transaction with `txid` `dc01120a639283c11eeada6ce540178adfbbca7fe0d2f41076367cbc9e901822`?
+
+1. Are there any places where a (w)txid is returned as a `uint256` where it would not make sense to return a `transaction_identifier` instead?
+
+1. The [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) class already exists. How does it already enforce type correctness, and how is it different from the approach in this PR?
+
+1. Are there any places where `GenTxid` is used but using `transaction_identifier` would be better?
+
+### Code Review
+
+{:start="10"}
+1. How is `transaction_identifier` able to subclass `uint256`, given that, in C++, integers are types and not classes?
+
+1. Why does `transaction_identifier` subclass `uint256` instead of being a completely new type?
+
+1. `Txid` and `Wtxid` are typedefs for `transaction_identifier<has_witness>`. Do you see any reason for code to directly use the `transaction_identifier` type? Should that be forbidden, and if so, how?
+
+1. [`transaction_identifier::Compare`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R39) is defined to return a `int`, yet `reinterpret_cast<const uint256&>` is used in the return statement. Is this correct, and if so, why? Why `reinterpret_cast` instead of another cast, such as `static_cast`?
```

Comment:
```suggestion
1. [`transaction_identifier::Compare`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R39) is defined to return an `int`, yet `reinterpret_cast<const uint256&>` is used in the return statement. Is this correct, and if so, why? Why `reinterpret_cast` instead of another cast, such as `static_cast`?
```